neural_structured_learning/examples/notebooks/adversarial_cnn_transfer_learning_fashionmnist.ipynb

```diff
@@ -0,0 +1,1263 @@
+        "# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
+        "# See the License for the specific language governing permissions and\n",
+        "# limitations under the License."
+      ],
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "85lXAvQDz7K-"
+      },
+      "source": [
+        "# Adversarial Learning: Building Robust Image Classifiers\n",
+        "\n",
+        "\n",
+        "## Overview\n",
+        "\n",
+        "In this tutorial, we will explore the use of adversarial learning\n",
+        "([Goodfellow et al., 2014](https://arxiv.org/abs/1412.6572)) for image\n",
+        "classification using Neural Structured Learning (NSL).\n",
+        "\n",
+        "Adversarial attacks intentionally introduce some noise in the form of perturbations to input images to fool the deep learning model. For example, in a classification system, by adding an imperceptibly small vector whose elements are equal to the sign of the elements of the gradient of the loss function with respect to the input, we can change the model's classification of the image. \n",
+        "\n",
+        "\n",
+        "## CNN Classifier\n",
+        "\n",
+        "The most popular deep learning models leveraged for computer vision problems are convolutional neural networks (CNNs)!\n",
+        "\n",
+        "![](https://i.imgur.com/32WEbHg.png)\n",
+        "<font size=2>Created by: Dipanjan Sarkar</font>\n",
+        "\n",
+        "We will look at how we can build, train and evaluate a multi-class CNN classifier in this notebook and also perform adversarial learning.\n",
+        "\n",
+        "\n",
+        "## Transfer Learning\n",
+        "\n",
+        "The idea is to leverage a pre-trained model instead of building a CNN from scratch in our image classification problem\n",
+        "\n",
+        "![](https://i.imgur.com/WcUabml.png)\n",
+        "<font size=2>Source: [CNN Essentials](https://github.com/dipanjanS/convolutional_neural_networks_essentials/tree/master/presentation)</font>\n",
+        "\n",
+        "## Tutorial Outline\n",
+        "\n",
+        "In this tutorial, we illustrate the following procedure of applying adversarial learning to obtain robust models using the Neural Structured Learning framework on a CNN model:\n",
+        "\n",
+        "1.  Create a neural network as a base model. In this tutorial, the base model is\n",
+        "    created with the `tf.keras` sequential API by wrapping a pre-trained `VGG19` model which we use for fine-tuning using transfer learning\n",
+        "2. Train and evaluate the base model performance on organic FashionMNIST data\n",
+        "3. Perform perturbations using the fast gradient sign method (FSGM) technique and look at model weaknesses\n",
+        "4. Generate perturbed dataset from the test data using FGSM and evaluate base model performance\n",
```

Comment:
I think this is actually being done towards the end, along with evaluating the adversarially trained model on the perturbed dataset. So, you may want to move this down accordingly.

Reply:
sure done